### PR TITLE
Add query 13 to TpchQueryBuilder

### PR DESCRIPTION
**Pull request by majetideepak**
Adds TPC-H query 13 to TpchQueryBuilder. 

Output of printPlanWithStats for TPC-H SF 10 using DWRF.
```
Execution time: 2.87s
Splits total: 160, finished: 160
-> OrderBy[custdist DESC NULLS LAST, c_count DESC NULLS LAST]
   Output: 46 rows (896B), Cpu time: 40.12us, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 1
  -> Aggregation[SINGLE [c_count] custdist := count(0)]
     Output: 46 rows (23.84KB), Cpu time: 7.04ms, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 1
    -> Aggregation[FINAL [c_custkey] c_count := count(ROW["pc_count"])]
       Output: 1500000 rows (34.29MB), Cpu time: 1.04s, Blocked wall time: 0ns, Peak memory: 56.00MB, Threads: 1
      -> LocalPartition[GATHER]
         Output: 24704006 rows (565.79MB), Cpu time: 51.88ms, Blocked wall time: 3.77s, Peak memory: 0B
         LocalPartition: Input: 12352003 rows (282.89MB), Output: 12352003 rows (282.89MB), Cpu time: 20.15ms, Blocked wall time: 1.92s, Peak memory: 0B, Threads: 8
         LocalExchange: Input: 12352003 rows (282.89MB), Output: 12352003 rows (282.89MB), Cpu time: 31.73ms, Blocked wall time: 1.85s, Peak memory: 0B, Threads: 1
        -> Aggregation[PARTIAL [c_custkey] pc_count := count(ROW["o_orderkey"])]
           Output: 12352003 rows (282.89MB), Cpu time: 1.95s, Blocked wall time: 0ns, Peak memory: 32.00MB, Threads: 8
          -> HashJoin[RIGHT o_custkey=c_custkey]
             Output: 15337604 rows (323.18MB), Cpu time: 1.95s, Blocked wall time: 480.61ms, Peak memory: 17.00MB
             HashBuild: Input: 1500000 rows (0B), Output: 0 rows (0B), Cpu time: 183.88ms, Blocked wall time: 42.31ms, Peak memory: 16.00MB, Threads: 8
             HashProbe: Input: 14837583 rows (972.32MB), Output: 15337604 rows (323.18MB), Cpu time: 1.77s, Blocked wall time: 438.30ms, Peak memory: 1.00MB, Threads: 8
            -> TableScan[table: orders, remaining filter: (not(like(ROW["comment"],"%special%requests%")))]
               Input: 14837583 rows (1.17GB), Raw Input: 15000000 rows (315.15MB), Output: 14837583 rows (972.32MB), Cpu time: 16.86s, Blocked wall time: 265.00us, Peak memory: 36.00MB, Threads: 8, Splits: 80
            -> TableScan[table: customer]
               Input: 1500000 rows (0B), Output: 1500000 rows (0B), Cpu time: 57.15ms, Blocked wall time: 389.00us, Peak memory: 3.00MB, Threads: 8, Splits: 80
```

### CODE REVIEW

File: velox/duckdb/conversion/DuckParser.cpp [hunk 1]
```diff
@@ -151,10 +152,16 @@ std::shared_ptr<const core::IExpr> parseFunctionExpr(ParsedExpression& expr) {
   for (const auto& c : functionExpr.children) {
     params.emplace_back(parseExpr(*c));
   }
-  return callExpr(
-      normalizeFuncName(functionExpr.function_name),
-      std::move(params),
-      getAlias(expr));
+  auto func = normalizeFuncName(functionExpr.function_name);
+  // NOT LIKE function needs special handling as it maps to two functions "not",
```
Comment (mbasmanova) [inline velox/duckdb/conversion/DuckParser.cpp:156]: Would you extract this change into a separate PR and add a test to DuckParserTest.cpp?

File: velox/dwio/parquet/reader/ParquetReader.cpp [hunk 2]
```diff
@@ -50,6 +50,24 @@ std::unique_ptr<::duckdb::ConstantFilter> constantEqualFilter(
       ::duckdb::ExpressionType::COMPARE_EQUAL, std::move(value));
 }
 
+void buildConjunctOrFilter(
```
Comment (mbasmanova) [inline velox/dwio/parquet/reader/ParquetReader.cpp:53]: It would be nice to extract this change into a separate PR and add a test to velox/dwio/parquet/tests/ParquetReaderTest.cpp

File: velox/exec/tests/utils/TpchQueryBuilder.cpp [hunk 3]
```diff
@@ -212,6 +214,70 @@ TpchPlan TpchQueryBuilder::getQ6Plan() const {
...
+                       {"c_custkey"},
+                       {"o_custkey"},
+                       notSpecialRequestOrders,
+                       "",
+                       {"c_custkey", "o_orderkey"},
+                       core::JoinType::kLeft)
+                   .partialAggregation({0}, {"count(o_orderkey) as pc_count"})
+                   .planNode()})
+          .finalAggregation({0}, {"count(pc_count) as c_count"}, {BIGINT()})
+          .partialAggregation({1}, {"count(0) as p_custdist"})
```
Comment (mbasmanova) [inline velox/exec/tests/utils/TpchQueryBuilder.cpp:263]: Since we already run single-threaded, this can be singleAggregation.

**Review by mbasmanova — COMMENTED** [review]
@majetideepak Thank you for adding q13 to the benchmark. I'm curious what is the performance of that query? Would you share the output of printPlanWithStats?

PR description is missing "13" in "Adds TPC-H query to TpchQueryBuilder."

File: velox/duckdb/conversion/DuckParser.cpp (hunk 1 above)
Comment (majetideepak) [reply velox/duckdb/conversion/DuckParser.cpp:156]: https://github.com/facebookincubator/velox/pull/1473

File: velox/dwio/parquet/reader/ParquetReader.cpp (hunk 2 above)
Comment (majetideepak) [reply velox/dwio/parquet/reader/ParquetReader.cpp:53]: https://github.com/facebookincubator/velox/pull/1472

**Review by mbasmanova — APPROVED** [review]
@majetideepak Thank you, Deepak. I'm super curious in the output of printPlanWithStats for q18 and how its performance compares to DuckDB.

**Comment by majetideepak** [comment]
@mbasmanova I am comparing the plans and results with DuckDB now. I will update here shortly.

**Comment by majetideepak** [comment]
@mbasmanova I changed the join order to make the build smaller than the probe. DuckDB is doing the same.
For tpch SF 10, Velox with DWRF took around 2.97s vs 1.92s for DuckDB on Parquet.
duckdbq13

**Comment by majetideepak** [comment]
@mbasmanova please re-import this since I changed the join order. Thanks.

**Comment by mbasmanova** [comment]
@majetideepak Thank you, Deepak, for sharing the profile. I see that final agg runs single-threaded.

```
    -> Aggregation[FINAL [c_custkey] c_count := count(ROW["pc_count"])]
       Output: 1500000 rows (34.29MB), Cpu time: 1.07s, Blocked wall time: 0ns, Peak memory: 56.00MB, Threads: 1
```

Try running it multi-threaded by adding localPartition({"c_custkey}) after partial agg.

**Comment by mbasmanova** [comment]
@majetideepak I'd also try breaking down SINGLE second agg into partial (multi-threaded) and final (single-threaded):

```
  -> Aggregation[SINGLE [c_count] custdist := count(0)]
     Output: 46 rows (23.84KB), Cpu time: 7.37ms, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 1
```

**Comment by majetideepak** [comment]
@mbasmanova I have made the changes you suggested. The final agg is still single-threaded.  Can you verify my change? thanks.
```
Execution time: 2.91s
Splits total: 160, finished: 160
-> OrderBy[custdist DESC NULLS LAST, c_count DESC NULLS LAST]
   Output: 46 rows (896B), Cpu time: 38.86us, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 1
  -> Aggregation[FINAL [c_count] custdist := count(ROW["p_custdist"])]
     Output: 46 rows (23.84KB), Cpu time: 67.96us, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 1
    -> Aggregation[PARTIAL [c_count] p_custdist := count(0)]
       Output: 46 rows (23.84KB), Cpu time: 7.38ms, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 1
      -> Aggregation[FINAL [c_custkey] c_count := count(ROW["pc_count"])]
         Output: 1500000 rows (34.29MB), Cpu time: 1.02s, Blocked wall time: 0ns, Peak memory: 56.00MB, Threads: 1
        -> LocalPartition[REPARTITION]
           Output: 24704006 rows (565.74MB), Cpu time: 56.80ms, Blocked wall time: 4.58s, Peak memory: 0B
           LocalPartition: Input: 12352003 rows (282.87MB), Output: 12352003 rows (282.87MB), Cpu time: 22.22ms, Blocked wall time: 2.68s, Peak memory: 0B, Threads: 8
           LocalExchange: Input: 12352003 rows (282.87MB), Output: 12352003 rows (282.87MB), Cpu time: 34.58ms, Blocked wall time: 1.90s, Peak memory: 0B, Threads: 1
          -> Aggregation[PARTIAL [c_custkey] pc_count := count(ROW["o_orderkey"])]
             Output: 12352003 rows (282.87MB), Cpu time: 1.88s, Blocked wall time: 0ns, Peak memory: 32.00MB, Threads: 8
            -> HashJoin[RIGHT o_custkey=c_custkey]
               Output: 15337604 rows (323.18MB), Cpu time: 1.96s, Blocked wall time: 462.36ms, Peak memory: 17.00MB
               HashBuild: Input: 1500000 rows (0B), Output: 0 rows (0B), Cpu time: 183.92ms, Blocked wall time: 37.02ms, Peak memory: 16.00MB, Threads: 8
               HashProbe: Input: 14837583 rows (972.32MB), Output: 15337604 rows (323.18MB), Cpu time: 1.78s, Blocked wall time: 425.35ms, Peak memory: 1.00MB, Threads: 8
              -> TableScan[Table: orders]
                 Input: 14837583 rows (1.17GB), Raw Input: 15000000 rows (315.15MB), Output: 14837583 rows (972.32MB), Cpu time: 16.70s, Blocked wall time: 552.00us, Peak memory: 36.00MB, Threads: 8, Splits: 80
              -> TableScan[Table: customer]
                 Input: 1500000 rows (0B), Output: 1500000 rows (0B), Cpu time: 55.55ms, Blocked wall time: 254.00us, Peak memory: 3.00MB, Threads: 8, Splits: 80

```

File: velox/exec/tests/utils/TpchQueryBuilder.cpp [hunk 4]
```diff
@@ -195,6 +197,65 @@ TpchPlan TpchQueryBuilder::getQ6Plan() const {
...
+                               "",
+                               {"c_custkey", "o_orderkey"},
+                               core::JoinType::kRight)
+                           .partialAggregation(
+                               {"c_custkey"}, {"count(o_orderkey) as pc_count"})
+                           .planNode()})
+                  .finalAggregation(
+                      {"c_custkey"}, {"count(pc_count) as c_count"}, {BIGINT()})
+                  .partialAggregation({"c_count"}, {"count(0) as p_custdist"})
+                  .finalAggregation({"c_count"}, {"count(p_custdist) as custdist"}, {BIGINT()})
```
Comment (mbasmanova) [inline velox/exec/tests/utils/TpchQueryBuilder.cpp:247]: You need a local exchange (gather) before this final aggregation.

**Comment by majetideepak** [comment]
@mbasmanova I see more threads now, but the wall time is slower. I see a lot of time spent blocked.

```
Execution time: 7.77s
Splits total: 160, finished: 160
-> OrderBy[custdist DESC NULLS LAST, c_count DESC NULLS LAST]
   Output: 46 rows (896B), Cpu time: 76.21us, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 1
  -> Aggregation[FINAL [c_count] custdist := count(ROW["p_custdist"])]
     Output: 46 rows (23.84KB), Cpu time: 152.29us, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 1
    -> LocalPartition[GATHER]
       Output: 670 rows (381.50KB), Cpu time: 8.32ms, Blocked wall time: 8.02s, Peak memory: 0B
       LocalPartition: Input: 335 rows (190.75KB), Output: 335 rows (190.75KB), Cpu time: 8.18ms, Blocked wall time: 49.38ms, Peak memory: 0B, Threads: 8
       LocalExchange: Input: 335 rows (190.75KB), Output: 335 rows (190.75KB), Cpu time: 132.91us, Blocked wall time: 7.97s, Peak memory: 0B, Threads: 1
      -> Aggregation[PARTIAL [c_count] p_custdist := count(0)]
         Output: 335 rows (190.75KB), Cpu time: 20.56ms, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 8
        -> Aggregation[FINAL [c_custkey] c_count := count(ROW["pc_count"])]
           Output: 1500000 rows (34.38MB), Cpu time: 2.13s, Blocked wall time: 0ns, Peak memory: 7.00MB, Threads: 8
          -> LocalPartition[REPARTITION]
             Output: 24704006 rows (565.65MB), Cpu time: 926.67ms, Blocked wall time: 59.11s, Peak memory: 6.00MB
             LocalPartition: Input: 12352003 rows (282.87MB), Output: 12352003 rows (282.87MB), Cpu time: 753.79ms, Blocked wall time: 2.89s, Peak memory: 6.00MB, Threads: 8
             LocalExchange: Input: 12352003 rows (282.78MB), Output: 12352003 rows (282.78MB), Cpu time: 172.87ms, Blocked wall time: 56.23s, Peak memory: 0B, Threads: 8
            -> Aggregation[PARTIAL [c_custkey] pc_count := count(ROW["o_orderkey"])]
               Output: 12352003 rows (282.87MB), Cpu time: 2.61s, Blocked wall time: 0ns, Peak memory: 28.00MB, Threads: 8
              -> HashJoin[RIGHT o_custkey=c_custkey]
                 Output: 15337604 rows (323.18MB), Cpu time: 2.79s, Blocked wall time: 725.22ms, Peak memory: 17.00MB
                 HashBuild: Input: 1500000 rows (0B), Output: 0 rows (0B), Cpu time: 253.67ms, Blocked wall time: 61.45ms, Peak memory: 16.00MB, Threads: 8
                 HashProbe: Input: 14837583 rows (972.32MB), Output: 15337604 rows (323.18MB), Cpu time: 2.53s, Blocked wall time: 663.78ms, Peak memory: 1.00MB, Threads: 8
                -> TableScan[Table: orders]
                   Input: 14837583 rows (1.17GB), Raw Input: 15000000 rows (315.15MB), Output: 14837583 rows (972.32MB), Cpu time: 23.88s, Blocked wall time: 88.00us, Peak memory: 36.00MB, Threads: 8, Splits: 80
                -> TableScan[Table: customer]
                   Input: 1500000 rows (0B), Output: 1500000 rows (0B), Cpu time: 62.38ms, Blocked wall time: 287.00us, Peak memory: 3.00MB, Threads: 8, Splits: 80

```

**Comment by mbasmanova** [comment]
> I see more threads now, but the wall time is slower. I see a lot of time spent blocked.

Looks like there might be some bugs. This would need to be investigated. One option is to go back to the original plan, land that, file an issue to optimize.

**Comment by majetideepak** [comment]
@mbasmanova I reverted to the plan with the best execution. This plan is also the same as the DuckDB one.
